src/Plugin/Action/GenerateImageDerivative.php

```diff
@@ -0,0 +1,218 @@
+
+  protected $mediaSource;
+
+  /**
+   * Constructs a EmitEvent action.
+   *
+   * @param array $configuration
+   *   A configuration array containing information about the plugin instance.
+   * @param string $plugin_id
+   *   The plugin_id for the plugin instance.
+   * @param mixed $plugin_definition
+   *   The plugin implementation definition.
+   * @param \Drupal\Core\Session\AccountInterface $account
+   *   Current user.
+   * @param \Drupal\Core\Entity\EntityTypeManager $entity_type_manager
+   *   Entity type manager.
+   * @param \Drupal\islandora\EventGenerator\EventGeneratorInterface $event_generator
+   *   EventGenerator service to serialize AS2 events.
+   * @param \Stomp\StatefulStomp $stomp
+   *   Stomp client.
+   * @param \Drupal\jwt\Authentication\Provider\JwtAuth $auth
+   *   JWT Auth client.
+   * @param \Drupal\islandora\IslandoraUtils $utils
+   *   Islandora utility functions.
+   */
+  public function __construct(
+    array $configuration,
+    $plugin_id,
+    $plugin_definition,
+    AccountInterface $account,
+    EntityTypeManager $entity_type_manager,
+    EventGeneratorInterface $event_generator,
+    StatefulStomp $stomp,
+    JwtAuth $auth,
+    IslandoraUtils $utils,
+    MediaSourceService $media_source
+  ) {
+    parent::__construct(
+      $configuration,
+      $plugin_id,
+      $plugin_definition,
+      $account,
+      $entity_type_manager,
+      $event_generator,
+      $stomp,
+      $auth
+    );
+    $this->utils = $utils;
+    $this->mediaSource = $media_source;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition) {
+    return new static(
+      $configuration,
+      $plugin_id,
+      $plugin_definition,
+      $container->get('current_user'),
+      $container->get('entity_type.manager'),
+      $container->get('islandora.eventgenerator'),
+      $container->get('islandora.stomp'),
+      $container->get('jwt.authentication.jwt'),
+      $container->get('islandora.utils'),
+      $container->get('islandora.media_source_service')
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function defaultConfiguration() {
+    return [
+      'queue' => 'islandora-connector-houdini',
+      'event' => 'Generate Derivative',
+      'source_term_uri' => '',
+      'derivative_term_uri' => '',
+      'mimetype' => 'image/jpeg',
+      'args' => '',
+    ];
+  }
+
+  protected function generateData($entity) {
+    $data = parent::generateData($entity);
+    
+    // Find media belonging to node that has the source term, and get its file url.
+    $source_term = $this->utils->getTermForUri($this->configuration['source_term_uri']);
+    $source_media = $this->utils->getMediaWithTerm($entity, $source_term);
+    $source_field = $this->mediaSource->getSourceFieldName($source_media->bundle());
+    $files = $source_media->get($source_field)->referencedEntities();
+    $file = reset($files);
```

Comment:
Field references always confuse me a bit but does this mean each field can only reference a single entity? If so, is this a choice or no other alternatives worked and lastly if so, we should document it so when people create their own objects they don't try to link multiple objects to the same field.

Reply:
Yeah, it's a gross PR because of features and shuffling around / deleting tons of config.  But you've both zeroed in on the right parts.

This bit in particular pulls out the file that the media is representing.  That's why I'm assuming there's only one.  This is pretty much the only place where that assumption is made, though.  Generally speaking, any entity reference can have multiple values at pretty much any point in time.

Reply:
@DiegoPino If you wanna see where all that jazz happens, it's in entity hooks in islandora.module: https://github.com/dannylamb/islandora/blob/content-modeling-overhaul/islandora.module#L98

There's a few layers, but basically, in the hook you check to see if you should trigger derivitves at all (e.g. has the file actually changed?).  If so it uses Context to evaluate conditions the user provides (e.g. Was it a Media that's a Preservation Master who belongs to an Image node that got updated?)  Then it fires a ContextReaction that's a small wrapper around core Drupal actions like our indexing and un-indexing.

The actual generating of the message and publishing it to the queue is what this Action does.

Reply:
Oh ok, so once you have a media entity we are assuming that each media entity references only one file? I'm okay with that assumption.